File: typescript/examples/io-write-atomic.test.ts
// カード io-write-atomic の Contract を検証するテスト
import { randomUUID } from 'node:crypto';
import { mkdir, mkdtemp, readFile, rename, rm, stat, unlink, writeFile } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { afterEach, beforeEach, describe, expect, it } from 'vitest';

let dir: string;

async function writeAtomic(target: string, data: string): Promise<void> {
  // 同じプロセス内で同じ対象へ同時に書いても衝突しないよう、呼び出しごとに一意な名前にする
  const tmp = `${target}.${process.pid}.${randomUUID()}.tmp`;
  try {
    await writeFile(tmp, data);
    await rename(tmp, target);
  } catch (err) {
    await unlink(tmp).catch(() => {}); // 失敗しても一時ファイルを残さない
    throw err;
  }
}

describe('io-write-atomic: rename', () => {
  beforeEach(async () => {
    dir = await mkdtemp(join(tmpdir(), 'io-write-atomic-'));
  });

  afterEach(async () => {
    await rm(dir, { recursive: true, force: true });
  });

  it('既存ファイルを上書きし、一時ファイルは消える', async () => {
    const target = join(dir, 'config.json');
    await writeFile(target, 'old');
    const tmp = join(dir, 'config.json.tmp');
    await writeFile(tmp, 'new');
    await expect(rename(tmp, target)).resolves.toBeUndefined();
    expect(await readFile(target, 'utf8')).toBe('new');
    await expect(stat(tmp)).rejects.toMatchObject({ code: 'ENOENT' });
  });

  it('置き換え後は一時ファイルの inode とパーミッションになる', async () => {
    const target = join(dir, 'secret.txt');
    await writeFile(target, 'old', { mode: 0o600 });
    const tmp = join(dir, 'secret.txt.tmp');
    await writeFile(tmp, 'new', { mode: 0o644 });
    const tmpStat = await stat(tmp);
    await rename(tmp, target);
    const after = await stat(target);
    expect(after.ino).toBe(tmpStat.ino);
    expect(after.mode & 0o777).toBe(0o644);
  });

  it('同時に読んでも途中状態（古い内容と新しい内容の混在）を見ない', async () => {
    const target = join(dir, 'data.txt');
    const a = 'A'.repeat(64 * 1024);
    const b = 'B'.repeat(64 * 1024);
    await writeFile(target, a);
    const writes = (async () => {
      for (let i = 0; i < 50; i++) await writeAtomic(target, i % 2 === 0 ? b : a);
    })();
    const reads = (async () => {
      const seen = new Set<string>();
      for (let i = 0; i < 50; i++) seen.add(await readFile(target, 'utf8'));
      return seen;
    })();
    const [, seen] = await Promise.all([writes, reads]);
    for (const content of seen) expect(content === a || content === b).toBe(true);
  });

  it('rename が失敗しても target は変わらず一時ファイルは残る', async () => {
    const target = join(dir, 'as-dir');
    await mkdir(target); // 置き換え先がディレクトリ
    const tmp = join(dir, 'as-dir.tmp');
    await writeFile(tmp, 'new');
    await expect(rename(tmp, target)).rejects.toMatchObject({ code: 'EISDIR' });
    expect((await stat(target)).isDirectory()).toBe(true);
    expect(await readFile(tmp, 'utf8')).toBe('new');
  });

  it('oldPath や newPath の親ディレクトリが無ければ ENOENT', async () => {
    const target = join(dir, 'x.txt');
    await expect(rename(join(dir, 'missing.tmp'), target)).rejects.toMatchObject({ code: 'ENOENT' });
    await writeFile(target, 'x');
    await expect(rename(target, join(dir, 'no-such-dir', 'x.txt'))).rejects.toMatchObject({ code: 'ENOENT' });
  });
});

File: python/examples/io-write-atomic_test.py
"""io-write-atomic: 一時ファイル + os.replace の Contract を検証する。"""

import errno
import os
import shutil
import sys
import tempfile
from pathlib import Path

import pytest


def write_atomic(target: Path, data: str) -> str:
    """カードの Usage と同じ手順。一時ファイル名を返す。"""
    with tempfile.NamedTemporaryFile("w", dir=target.parent, delete=False, encoding="utf-8") as tmp:
        tmp_name = tmp.name
        try:
            tmp.write(data)
            tmp.flush()
            os.fsync(tmp.fileno())
        except BaseException:
            Path(tmp_name).unlink(missing_ok=True)
            raise
    try:
        os.replace(tmp_name, target)
    except BaseException:
        # replace に失敗したら一時ファイルを残さない（別デバイスの EXDEV など）
        Path(tmp_name).unlink(missing_ok=True)
        raise
    return tmp_name


def test_replaces_existing_file_and_removes_tmp(tmp_path: Path) -> None:
    """既存ファイルを上書きし、一時ファイルは無くなり、dst は一時ファイルの inode になる。"""
    target = tmp_path / "config.json"
    target.write_text("old", encoding="utf-8")
    src = tmp_path / "src"
    src.write_text("new", encoding="utf-8")
    ino = os.stat(src).st_ino
    assert os.replace(src, target) is None
    assert target.read_text(encoding="utf-8") == "new"
    assert not src.exists()
    assert os.stat(target).st_ino == ino


def test_usage_idiom_creates_or_replaces(tmp_path: Path) -> None:
    """Usage の手順で新規作成も置き換えもでき、一時ファイルは残らない。"""
    target = tmp_path / "config.json"
    tmp1 = write_atomic(target, '{"n": 1}')
    tmp2 = write_atomic(target, '{"n": 2}')
    assert target.read_text(encoding="utf-8") == '{"n": 2}'
    assert not os.path.exists(tmp1) and not os.path.exists(tmp2)
    assert sorted(p.name for p in tmp_path.iterdir()) == ["config.json"]


def test_reader_opened_before_replace_sees_old_content(tmp_path: Path) -> None:
    """置き換え前から開いていたファイルオブジェクトは旧内容を読み続ける。"""
    target = tmp_path / "t.txt"
    target.write_text("old", encoding="utf-8")
    with open(target, encoding="utf-8") as reader:
        write_atomic(target, "new")
        assert reader.read() == "old"
    assert target.read_text(encoding="utf-8") == "new"


def test_failure_during_write_leaves_target_and_tmp(tmp_path: Path) -> None:
    """書いている途中で例外が出ても dst は旧内容のままで、一時ファイルは残る。"""
    target = tmp_path / "t.txt"
    target.write_text("old", encoding="utf-8")
    with pytest.raises(RuntimeError):
        with tempfile.NamedTemporaryFile("w", dir=tmp_path, delete=False, encoding="utf-8") as tmp:
            tmp.write("par")
            raise RuntimeError("boom")
    assert target.read_text(encoding="utf-8") == "old"
    assert tmp.closed
    assert Path(tmp.name).read_text(encoding="utf-8") == "par"


def test_error_cases(tmp_path: Path) -> None:
    """src 無し・dst の親無しは FileNotFoundError、dst がディレクトリは IsADirectoryError など。"""
    f = tmp_path / "f"
    f.write_text("x", encoding="utf-8")
    d = tmp_path / "d"
    d.mkdir()
    with pytest.raises(FileNotFoundError):
        os.replace(tmp_path / "nope", f)
    with pytest.raises(FileNotFoundError):
        os.replace(f, tmp_path / "nodir" / "x")
    with pytest.raises(IsADirectoryError):
        os.replace(f, d)
    with pytest.raises(NotADirectoryError):
        os.replace(d, f)
    nonempty = tmp_path / "ne"
    nonempty.mkdir()
    (nonempty / "child").write_text("", encoding="utf-8")
    with pytest.raises(OSError) as info:
        os.replace(d, nonempty)
    assert info.value.errno in (errno.ENOTEMPTY, errno.EEXIST)
    assert f.read_text(encoding="utf-8") == "x"


@pytest.mark.skipif(sys.platform == "win32", reason="POSIX のパーミッション")
def test_permission_comes_from_tmp_file(tmp_path: Path) -> None:
    """パーミッションは一時ファイルの 0600 になる。copymode で引き継げる。"""
    target = tmp_path / "t.txt"
    target.write_text("old", encoding="utf-8")
    os.chmod(target, 0o644)
    write_atomic(target, "new")
    assert os.stat(target).st_mode & 0o777 == 0o600

    os.chmod(target, 0o644)
    with tempfile.NamedTemporaryFile("w", dir=tmp_path, delete=False, encoding="utf-8") as tmp:
        tmp.write("new2")
    shutil.copymode(target, tmp.name)
    os.replace(tmp.name, target)
    assert os.stat(target).st_mode & 0o777 == 0o644


def _other_device_dir(base: Path) -> Path | None:
    """base と st_dev が異なるディレクトリを探す（無ければ None）。"""
    dev = os.stat(base).st_dev
    for cand in ("/dev/shm", "/run", "/System/Volumes/VM"):
        try:
            if os.stat(cand).st_dev != dev:
                return Path(cand)
        except OSError:
            continue
    return None


def test_cross_device_raises_exdev(tmp_path: Path) -> None:
    """別デバイスへの置き換えは OSError（errno.EXDEV）。"""
    other = _other_device_dir(tmp_path)
    if other is None:
        pytest.skip("別デバイスのディレクトリが見つからない")
    src = tmp_path / "src"
    src.write_text("x", encoding="utf-8")
    with pytest.raises(OSError) as info:
        os.replace(src, other / f"ai-repertoire-{os.getpid()}.tmp")
    assert info.value.errno == errno.EXDEV
    assert src.exists()


@pytest.mark.skipif(sys.platform == "win32", reason="Windows では os.rename が失敗する")
def test_os_rename_also_overwrites_on_posix(tmp_path: Path) -> None:
    """POSIX では os.rename も既存ファイルを上書きする（違いは Windows だけ）。"""
    a = tmp_path / "a"
    b = tmp_path / "b"
    a.write_text("A", encoding="utf-8")
    b.write_text("B", encoding="utf-8")
    os.rename(a, b)
    assert b.read_text(encoding="utf-8") == "A"
